File: richos/mobile/native-ios/Core/Sources/RichOSCore/Protocol/Signing.swift
import CryptoKit
import Foundation

/// Request signing, phone protocol contract §3 — checked case by case against
/// `richos/mobile/conformance/vectors/signing.json`, whose signatures the Mac's own verifier accepts.
///
/// The traps, each a conformance case: the fourth line is EMPTY when there is no body (not the hash
/// of zero bytes); the path is signed as sent on the wire (percent-encoding kept, `auth` removed);
/// the method is uppercased; the signature is RAW 64-byte r||s in base64url without padding —
/// `SecKeyCreateSignature` and Android's `SHA256withECDSA` return DER and must be converted.
public enum RequestSigning {
    /// `<challenge>\n<METHOD>\n<path?query minus auth>\n<sha256 hex of the body, or empty>`.
    public static func canonicalString(challenge: String, method: String, pathWithQuery: String, body: Data?) -> String {
        [challenge, method.uppercased(), signedPath(pathWithQuery), bodyHashHex(body)].joined(separator: "\n")
    }

    /// Lowercase hex SHA-256 of the body bytes, or the empty string when there are none.
    public static func bodyHashHex(_ body: Data?) -> String {
        guard let body, !body.isEmpty else { return "" }
        return SHA256.hash(data: body).map { String(format: "%02x", $0) }.joined()
    }

    /// The path and query as sent, with the `auth` parameter removed and empty `&&` segments dropped
    /// (the Mac's `signed_path`, `phone/routes.rs`). Percent-encoding is never undone.
    public static func signedPath(_ target: String) -> String {
        guard let q = target.firstIndex(of: "?") else { return target }
        let path = String(target[..<q])
        let kept = target[target.index(after: q)...].split(separator: "&", omittingEmptySubsequences: true)
            .filter { !($0 == "auth" || $0.hasPrefix("auth=")) }
        return kept.isEmpty ? path : path + "?" + kept.joined(separator: "&")
    }

    /// `RichOS-Device <device_id>.<challenge>.<signature_b64url>`.
    public static func authorization(deviceID: String, challenge: String, signature: Data) -> String {
        "RichOS-Device \(deviceID).\(challenge).\(Base64URL.encode(signature))"
    }

    /// The Mac splits the credential from the right (`phone/device.rs`).
    public static func parseAuthorization(_ value: String) -> (deviceID: String, challenge: String, signature: String)? {
        guard value.hasPrefix("RichOS-Device ") else { return nil }
        let rest = value.dropFirst("RichOS-Device ".count)
        guard let last = rest.lastIndex(of: ".") else { return nil }
        let signature = String(rest[rest.index(after: last)...])
        let head = rest[..<last]
        guard let middle = head.lastIndex(of: ".") else { return nil }
        return (String(head[..<middle]), String(head[head.index(after: middle)...]), signature)
    }

    /// `GET /api/events` carries the credential as the LAST query parameter, `auth=<encoded>`
    /// (contract §3.2); a header on that route is ignored.
    public static func withQueryCredential(_ pathWithQuery: String, authorization: String) -> String {
        pathWithQuery + (pathWithQuery.contains("?") ? "&" : "?") + "auth=" + encodeURIComponent(authorization)
    }

    /// JavaScript's `encodeURIComponent`, which the reference client uses for every query value.
    public static func encodeURIComponent(_ value: String) -> String {
        var allowed = CharacterSet.alphanumerics.intersection(CharacterSet(charactersIn: Unicode.Scalar(0)...Unicode.Scalar(127)))
        allowed.insert(charactersIn: "-_.!~*'()")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    /// A DER `ECDSA-Sig-Value` as the raw 64-byte r||s the Mac requires.
    public static func rawSignature(fromDER der: Data) throws -> Data {
        try P256.Signing.ECDSASignature(derRepresentation: der).rawRepresentation
    }
}

/// base64url without padding (RFC 4648 §5), the encoding of every key and signature on the wire.
public enum Base64URL {
    public static func encode(_ data: Data) -> String {
        data.base64EncodedString().replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    /// Accepts padded or unpadded input.
    public static func decode(_ text: String) -> Data? {
        var s = text.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        s = s.replacingOccurrences(of: "=", with: "")
        s += String(repeating: "=", count: (4 - s.count % 4) % 4)
        return Data(base64Encoded: s)
    }
}

/// This phone's identity toward the Mac, derived from its public key (contract §2.2).
public struct DeviceIdentity: Equatable, Sendable {
    /// The 65-byte uncompressed P-256 point.
    public var publicPoint: Data

    public init(publicPoint: Data) { self.publicPoint = publicPoint }

    /// `"dev_"` + lowercase hex of the first 6 bytes of SHA-256(point) — derived, never chosen.
    public var deviceID: String {
        "dev_" + SHA256.hash(data: publicPoint).prefix(6).map { String(format: "%02x", $0) }.joined()
    }

    /// `{"kty":"EC","crv":"P-256","x","y"}` with 32-byte coordinates in base64url.
    public var jwk: [String: String] {
        ["kty": "EC", "crv": "P-256",
         "x": Base64URL.encode(publicPoint.subdata(in: 1..<33)),
         "y": Base64URL.encode(publicPoint.subdata(in: 33..<65))]
    }
}

/// Signs canonical strings with this phone's key. The iPhone keeps a non-extractable key in the
/// Secure Enclave where it can (build plan §3.2); tests and the headless CLI use `SoftwareSigner`.
public protocol Signer: Sendable {
    func publicPoint() async throws -> Data
    /// Raw 64-byte r||s over SHA-256 of `message`.
    func sign(_ message: Data) async throws -> Data
}

/// A software P-256 key (CryptoKit). For tests, the headless CLI and the simulator.
public struct SoftwareSigner: Signer {
    private let key: P256.Signing.PrivateKey

    public init(key: P256.Signing.PrivateKey = .init()) { self.key = key }

    /// From a raw 32-byte scalar (the conformance corpus's published test key).
    public init(rawScalar: Data) throws { key = try P256.Signing.PrivateKey(rawRepresentation: rawScalar) }

    public func publicPoint() -> Data { key.publicKey.x963Representation }

    public func sign(_ message: Data) throws -> Data { try key.signature(for: message).rawRepresentation }
}

extension Signer {
    /// Signs one request and returns its credential. `pathWithQuery` is the wire form without `auth`.
    public func authorization(deviceID: String, challenge: String, method: String, pathWithQuery: String, body: Data?) async throws -> String {
        let canonical = RequestSigning.canonicalString(challenge: challenge, method: method, pathWithQuery: pathWithQuery, body: body)
        return RequestSigning.authorization(deviceID: deviceID, challenge: challenge, signature: try await sign(Data(canonical.utf8)))
    }
}

File: richos/mobile/native-android/core/src/main/kotlin/dev/richos/android/core/protocol/Signing.kt
package dev.richos.android.core.protocol

import dev.richos.android.core.CoreError
import java.net.URLEncoder
import java.security.MessageDigest
import java.util.Base64

/**
 * Request signing (phone protocol contract §3), as pure functions so every byte is proven on the
 * JVM. The key itself never appears here: the app signs with an Android Keystore key it cannot
 * export and hands the result to [derToRaw].
 */
object Signing {
    private val b64url = Base64.getUrlEncoder().withoutPadding()
    private val b64urlDecoder = Base64.getUrlDecoder()

    fun sha256(bytes: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(bytes)

    fun hex(bytes: ByteArray): String = bytes.joinToString("") { "%02x".format(it.toInt() and 0xff) }

    fun base64url(bytes: ByteArray): String = b64url.encodeToString(bytes)

    fun fromBase64url(text: String): ByteArray = b64urlDecoder.decode(text)

    /**
     * `challenge \n METHOD \n path-with-query \n body-hash` (§3.1). The path is the WIRE form,
     * percent-encoding exactly as sent, without the `auth` parameter. The fourth line is the
     * EMPTY string when there is no body — not the hash of zero bytes.
     */
    fun signingString(challenge: String, method: String, pathWithQuery: String, body: ByteArray?): String {
        val bodyLine = if (body == null || body.isEmpty()) "" else hex(sha256(body))
        return "$challenge\n${method.uppercase()}\n$pathWithQuery\n$bodyLine"
    }

    /** `RichOS-Device <device_id>.<challenge>.<signature base64url>` (§3.2). */
    fun authorization(deviceId: String, challenge: String, rawSignature: ByteArray): String {
        if (rawSignature.size != 64) throw CoreError("a P-256 signature is 64 raw bytes (r||s)")
        return "RichOS-Device $deviceId.$challenge.${base64url(rawSignature)}"
    }

    /**
     * The event stream carries its credential as the LAST query parameter, `auth`, percent-encoded
     * like JavaScript's `encodeURIComponent`; it is not part of the signed path (§3.2, §5.4).
     */
    fun withAuthQuery(signedPathWithQuery: String, authorization: String): String {
        val separator = if ('?' in signedPathWithQuery) '&' else '?'
        return "$signedPathWithQuery${separator}auth=${encodeURIComponent(authorization)}"
    }

    fun encodeURIComponent(text: String): String =
        URLEncoder.encode(text, Charsets.UTF_8)
            .replace("+", "%20")
            .replace("%21", "!").replace("%27", "'").replace("%28", "(").replace("%29", ")").replace("%7E", "~")

    /**
     * Android Keystore's `SHA256withECDSA` returns ASN.1 DER; the Mac verifies the raw 64-byte
     * r||s form (§0 item 2). SEQUENCE { INTEGER r, INTEGER s } → r and s, each left-padded to 32.
     */
    fun derToRaw(der: ByteArray): ByteArray {
        var i = 0
        fun byte(): Int = if (i < der.size) der[i++].toInt() and 0xff else throw CoreError("truncated DER signature")
        fun length(): Int {
            val first = byte()
            if (first < 0x80) return first
            val count = first and 0x7f
            if (count == 0 || count > 2) throw CoreError("unsupported DER length")
            var n = 0
            repeat(count) { n = (n shl 8) or byte() }
            return n
        }
        fun integer(): ByteArray {
            if (byte() != 0x02) throw CoreError("DER signature: expected an INTEGER")
            val len = length()
            if (len < 1 || i + len > der.size) throw CoreError("DER signature: bad INTEGER length")
            var value = der.copyOfRange(i, i + len)
            i += len
            while (value.size > 1 && value[0] == 0.toByte()) value = value.copyOfRange(1, value.size)
            if (value.size > 32) throw CoreError("DER signature: INTEGER wider than 32 bytes")
            return ByteArray(32 - value.size) + value
        }
        if (byte() != 0x30) throw CoreError("DER signature: expected a SEQUENCE")
        val total = length()
        if (i + total != der.size) throw CoreError("DER signature: length does not match")
        val r = integer()
        val s = integer()
        if (i != der.size) throw CoreError("DER signature: trailing bytes")
        return r + s
    }

    /** The inverse of [derToRaw], so a raw vector can be verified with the JVM's own verifier. */
    fun rawToDer(raw: ByteArray): ByteArray {
        if (raw.size != 64) throw CoreError("a raw P-256 signature is 64 bytes")
        fun integer(part: ByteArray): ByteArray {
            var v = part.dropWhile { it == 0.toByte() }.toByteArray()
            if (v.isEmpty()) v = byteArrayOf(0)
            if (v[0].toInt() and 0x80 != 0) v = byteArrayOf(0) + v
            return byteArrayOf(0x02, v.size.toByte()) + v
        }
        val body = integer(raw.copyOfRange(0, 32)) + integer(raw.copyOfRange(32, 64))
        return byteArrayOf(0x30, body.size.toByte()) + body
    }

    /** `"dev_"` + lowercase hex of the first 6 bytes of SHA-256(65-byte uncompressed point) (§2.2). */
    fun deviceId(uncompressedPoint: ByteArray): String {
        if (uncompressedPoint.size != 65 || uncompressedPoint[0] != 0x04.toByte()) throw CoreError("expected a 65-byte uncompressed P-256 point")
        return "dev_" + hex(sha256(uncompressedPoint).copyOfRange(0, 6))
    }

    /** The public half as the JWK the pairing request carries (§2.2). */
    fun publicJwk(uncompressedPoint: ByteArray): Map<String, String> {
        if (uncompressedPoint.size != 65 || uncompressedPoint[0] != 0x04.toByte()) throw CoreError("expected a 65-byte uncompressed P-256 point")
        return linkedMapOf(
            "kty" to "EC",
            "crv" to "P-256",
            "x" to base64url(uncompressedPoint.copyOfRange(1, 33)),
            "y" to base64url(uncompressedPoint.copyOfRange(33, 65)),
        )
    }
}
